### PR TITLE
Fix critical proposer selection bug #4259

### LINKED ISSUE #4259: ActiveValidatorIndices helper function returns wrong indices
The ActiveValidatorIndices function is supposed to return the indices of the currently active validators.
But when appending to the indices array it appends the index of the validator in the array returned by `activeIndicesCache.ActiveIndicesInEpoch(epoch)` and not the actual index of the validator.

See:
https://github.com/prysmaticlabs/prysm/blob/0326be86b5781b8e9d40d8f003d28a895e71203f/beacon-chain/core/helpers/validators.go#L82-L86

That is why a validator with an index greater than the number of active validators will never get any slot proposal or attestation assignments.

Unfortunately the type `v1alpha1.Validator` that is returned by `activeIndicesCache.ActiveIndicesInEpoch(epoch)` also does not contain the validator index.

### PR DESCRIPTION
**Pull request by prestonvanloon**
Resolves #4259.

This is likely consensus breaking in the current testnet. 

OK to merge into 0.9.2. Testnet restart required!

### CODE REVIEW

**Comment by terencechain** [comment]
Changed base to v0.9.2

I can help you resolve the conflict if needed

**Comment by prestonvanloon** [comment]
Is this confirmed consensus breaking?

Can you sync 0.9.2 with master? Why are there conflicts?

**Comment by terencechain** [comment]
> Is this confirmed consensus breaking?

yes, 2 nodes both have different proposer selection logic

**Comment by prestonvanloon** [comment]
Let's wait to change base until 0.9.2 is synced. Otherwise the diff is impossible to review

**Comment by prestonvanloon** [comment]
I did see this when trying to sync with these changes
```
[2019-12-11 15:35:40] DEBUG forkchoice: Executing state transition on block slot=115
panic: could not process blockCopy from fork choice service: could not execute state transition: could not process block: could not process block header: parent root 0xde64d3607af5b1ea9bf3cd8ff6fe0bc1baf8c27155
c61eebffa2385b747b4383 does not match the latest block header signing root in state 0xc170c1197d52b9a636c4919ddb8afd502b8685b26ef7412b189c3b17890662e2

goroutine 295 [running]:
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*InitialSync).Start(0xc00174c540)
        beacon-chain/sync/initial-sync/service.go:132 +0xf47
created by github.com/prysmaticlabs/prysm/shared.(*ServiceRegistry).StartAll
        shared/service_registry.go:44 +0x23e
```

**Comment by prestonvanloon** [comment]
Some other test is failing, trying to fix that.

**Comment by prestonvanloon** [comment]
`TestServer_ListAssignments_Pagination_DefaultPageSize_FromArchive` is failing. @rauljordan can you take a look when you have a moment?

**Comment by terencechain** [comment]
> What is this? Duplicate logic?

Duplicated logic for archival service. @rauljordan is it needed?

**Comment by terencechain** [comment]
This is ready. Open items:
1. Found there's duplicated proposer selection logic for archival service
2. E2e testing failure

**Comment by prestonvanloon** [comment]
This is failing e2e because 0.9.2 is failing e2e, unfortunately. So we don't know this this works at runtime.

**Comment by terencechain** [comment]
> This is failing e2e because 0.9.2 is failing e2e, unfortunately. So we don't know this this works at runtime.

v0.9.2 doesn't work run time
`sync<->pool<->rpc` is undergoing major refactor at this moment

**Comment by terencechain** [comment]
For this, we should add a regression test where pre computed `proposer_index` is greater than `len(active_vaidators)`,  I don't see there's one yet

**Comment by prestonvanloon** [comment]
I will refactor ComputeProposerIndex to accept a list of validators rather than the whole state.

```go
func ComputeProposerIndex(validators []*pb.Validator, indices []uint64, seed [32]byte) (uint64, error) {
```

This would allow deduplication. I'm boarding flight now, will push the changes when the flight lands.

**Comment by prestonvanloon** [comment]
Refactored for deduplication, added regression test in response to @terencechain feedback, added panic safety and appropriate test. 

This is ready for final review.

File: beacon-chain/core/helpers/validators_test.go [hunk 1]
```diff
@@ -405,3 +405,126 @@ func TestActiveValidatorIndices(t *testing.T) {
...
+		indices    []uint64
+		seed       [32]byte
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    uint64
+		wantErr bool
+	}{
+		{
```
Comment (nisdas) [inline beacon-chain/core/helpers/validators_test.go]: can you also one more test case that closely mimics for what we observed in our testnet ?  example: `validator > 600` isnt selected for proposal

Comment (prestonvanloon) [reply beacon-chain/core/helpers/validators_test.go]: There is one for that, I forgot to rename it, but see `second_half_active` which is the regression that only the first 5 would have been selected when the latter 5 were the only ones active.

**Pull request merged**